### PR TITLE
draft init rewrite

### LINKED ISSUE #82: RFC: adding built-in docker registry , do not dependent thirdparty docker registry
Is it possible to add a built-in docker registry inside kubernet cluster, no third party docker registry need. 

For enterprise usage, we often not allow to put our image outside or to the third party docker registry. 

So if we can set up a built-in docker registry inside our kubernete cluster , it will be very useful.

### LINKED ISSUE #102: Allow for full local development with minikube
When using minikube, we can re-use the docker daemon that is running on the VM without having to use an external registry, as stated in the documentation:
 
> When using a single VM of Kubernetes, it’s really handy to reuse the minikube’s built-in Docker daemon; as this means you don’t have to build a docker registry on your host machine and push the image into it - you can just build inside the same docker daemon as minikube which speeds up local experiments.

Would it be possible to replicate this experience with draft and document how to do it?

### LINKED ISSUE #122: Error: error when upgrading: rpc error: code = Unknown desc = YAML parse error on draftd/templates/deployment.yaml
Receiving the following error when trying to init or upgrade.

```
Error: error when upgrading: rpc error: code = Unknown desc = YAML parse error on draftd/templates/deployment.yaml: error converting YAML to JSON: yaml: line 25: could not find expected ':'
```

```
draft init --set registry.url=draftregistry.azurecr.io,registry.org=draftregistry,registry.authtoken=$authtoken,basedomain=draft.readon.ly --upgrade
Creating pack golang
Creating pack php
Creating pack node
Creating pack java
Creating pack ruby
Creating pack python
$DRAFT_HOME has been configured at /Users/bverst/.draft.
Error: error when upgrading: rpc error: code = Unknown desc = YAML parse error on draftd/templates/deployment.yaml: error converting YAML to JSON: yaml: line 25: could not find expected ':'
```

```
Client: &version.Version{SemVer:"canary", GitCommit:"ed93463b23f1ebb5ee6f670b6dbbb51df11aa46e", GitTreeState:"clean"}
Server: &version.Version{SemVer:"canary", GitCommit:"34f37fd12ce5d020867a40c8baeb83cb02286579", GitTreeState:"dirty"}
```

### LINKED ISSUE #152: Make draft init foolproof
Requiring users to copy/paste/change this line:

```
draft init --set registry.url=changeme,registry.org=changeme,registry.authtoken=changeme,basedomain=changeme
```

has caused some false starts and pain for first timers.

What could be done to make `draft init` be able to successfully install draft on a kubernetes cluster without any intense documentation parsing?

### LINKED ISSUE #167: draft init requires email, although it's not really needed by the OSS registry
When following: https://github.com/Azure/draft/blob/master/docs/install.md I see the command
```
$ echo '{"username":"jdoe","password":"secret","email":"jdoe@acme.com"}' | base64
```
Knowing registries don't actually require email, I just took it off. After a 'draft up' I got the error:
```Error encountered streaming JSON response: no basic auth credentials```
It took a while to even get this far, as CLIs don't apparently do auto-correct :(. So, it took a bit more to make sure the username/password combination were proper and just happen to try adding the email.
Based on the refactor, I see this encoding may go away, which would be perfect. But, I would imagine there'd be a CLI that still took --username --password, and just suggesting we remove any dependency for the user to pass in email. 
Was going to do a PR for the docs to note this requirement until the refactoring is done, but I believe the change is quickly pending.

### PR DESCRIPTION
**Pull request by bacongobbler**
closes #82
closes #102
closes #122
closes #152 
closes #167

This is a relatively "large" PR, so here's where we are at today:

- Users can no longer update draft configuration through `-f` and `--set`; user will be prompted for this information instead if required
- `--yes` is available if user wants to auto-accept the defaults
- If "minikube" is detected as the current context in ~/.kube/config, then we assume that they have installed the `ingress` and `registry` addons as well as ran `helm init` after the cluster was created. `draft init` detects this and drops some configuration to use these addons natively

TODO:

 - [x] prompt for missing information
 - [x] update "Getting started" documentation

### CODE REVIEW

**Comment by bacongobbler** [comment]
bit of a sample (with a custom minikube build with registry addon support):

```
$ minikube addons enable registry
$ minikube addons enable ingress
$ minikube start
$ helm init
$ draft init
Creating pack php...
Pack php already exists. Skipping!
Creating pack golang...
Pack golang already exists. Skipping!
Creating pack ruby...
Pack ruby already exists. Skipping!
Creating pack java...
Pack java already exists. Skipping!
Creating pack python...
Pack python already exists. Skipping!
Creating pack node...
Pack node already exists. Skipping!
$DRAFT_HOME has been configured at /Users/bacongobbler/.draft.
Draftd detected that you are using 'minikube' as your cloud provider. AWESOME!
Draftd is using the following configuration:

\``` [NOTE: backslashes added to escape triple backtick]
registry:
  url: $(REGISTRY_SERVICE_HOST)
  org: draft
basedomain: k8s.local
\```

Draftd has been installed into your Kubernetes Cluster.
Happy Sailing!
```

**Comment by bacongobbler** [comment]
testing this feature requires the following PR: https://github.com/kubernetes/minikube/pull/1583

If you don't want to build minikube from scratch, you can just run

```
minikube addons enable ingress
minikube start --insecure-registry 10.0.0.0/24
kubectl create -f deploy/addons/registry/registry-rc.yaml
kubectl create -f deploy/addons/registry/registry-svc.yaml
helm init
draft init
```

**Comment by bacongobbler** [comment]
closes #82

**Comment by bacongobbler** [comment]
testers of this PR will also require https://github.com/kubernetes/minikube/pull/1603. I accidentally broke the original PR on a last-minute fix.

**Comment by sgoings** [comment]
What is your acceptance criteria for getting this merged? Any support for GKE/AWS k8s clusters (besides minikube)?

Right now if the provider isn't identified the messaging can be kind of confusing:

```
Creating pack java...
Pack java already exists. Skipping!
Creating pack python...
Pack python already exists. Skipping!
Creating pack golang...
Pack golang already exists. Skipping!
Creating pack node...
Pack node already exists. Skipping!
Creating pack php...
Pack php already exists. Skipping!
Creating pack ruby...
Pack ruby already exists. Skipping!
$DRAFT_HOME has been configured at /Users/sgoings/.draft.

Draftd detected that you are using  as your cloud provider.
Draftd would be using the following configuration:

<code block>
<code block>

Is this okay? (Y/n)
```

**Comment by sgoings** [comment]
Also if I select "n" on that prompt, the ux doesn't feel quite right (draft moves on with changes instead of giving me cancel-like feedback):

```
Is this okay? (Y/n) n
Draftd has been installed into your Kubernetes Cluster.
Happy Sailing!
```

**Comment by bacongobbler** [comment]
Yeah right now it's just minikube. Acceptance criteria for this PR is:

- prompt for unknown cloud providers
- update docs to reference the new QuickStart docs

Azure support would be a nice one to get in, but we can add support for that in a follow-up and just roll with minikube for now.

**Comment by bacongobbler** [comment]
When you select "no" it should start to prompt you for manual information. That hasn't been implemented yet. Would that be ok to use or would you prefer a different UX?

**Comment by sgoings** [comment]
I like that approach. 

- prompt for unknown cloud providers
- a "no" starts prompting for manual information
- doc update

Ping me when those changes go in! I like where this is headed.

**Comment by squillace** [comment]
+1

**Comment by bacongobbler** [comment]
I've revised the installation guide to make minikube front and center, as I imagine a lot of our userbase will be running a local cluster during development rather than on a cloud provider. @sgoings I haven't changed the unknown cloud providers bit, but would you like to take a look at the installation guide and see if that works for you?

**Comment by bacongobbler** [comment]
Something to note: this PR cannot be merged until minikube v0.20 has been released as the `registry` add-on is a new feature in master.

File: chart/values.yaml [hunk 1]
```diff
@@ -28,4 +28,4 @@ registry:
   # For token-based logins, use
   #
   # $ echo '{"registrytoken":"9cbaf023786cd7"}' | base64
-  authtoken: changeme
+  authtoken: e30K
```
Comment (michelleN) [inline chart/values.yaml]: why this change? how come we need an auth token for pulling public images?

Comment (bacongobbler) [inline chart/values.yaml]: "e30K" is the base64 equivalent of "{}". It's actually a very weird docker behaviour in which you NEED to send an empty auth token even on registries with no authentication (like minikube w/ an in-cluster registry). Sending a string like "changeme" just causes docker to freak out because it can't parse it into a json object.

Comment (bacongobbler) [inline chart/values.yaml]: See https://github.com/moby/moby/blob/254fc83cba90ed79c78f4cb0cb33aeeaff492798/client/image_push.go#L54 as an example. Even if the registry isn't backed by auth, an empty auth token still gets shipped and the v2 registry expects that.

**Comment by bacongobbler** [comment]
k, so the documentation has been rewritten with minikube in mind and the new "prompt for missing information" dialogue is available for users running on Azure/GCS/AWS/whathaveyou.

Sample output:

```
><> draft init
Creating pack java...
Pack java already exists. Skipping!
Creating pack node...
Pack node already exists. Skipping!
Creating pack php...
Pack php already exists. Skipping!
Creating pack ruby...
Pack ruby already exists. Skipping!
Creating pack python...
Pack python already exists. Skipping!
Creating pack golang...
Pack golang already exists. Skipping!
$DRAFT_HOME has been configured at /Users/bacongobbler/.draft.

Draft detected that you are using minikube as your cloud provider. AWESOME!
Draft will be using the following configuration:

'''
registry:
  url: $(REGISTRY_SERVICE_HOST)
basedomain: k8s.local
'''

Is this okay? (Y/n) nah

In order to install Draft, we need a bit more information...

1. Enter your Docker registry URL (e.g. docker.io, quay.io, myregistry.azurecr.io): docker.io
2. Enter your username: bacongobbler
3. Enter your password: 
4. Enter your org (bacongobbler): 
5. Enter the top-level domain: draft.bacongobbler.com
Warning: Draft is already installed in the cluster.
Use --client-only to suppress this message.
Happy Sailing!
```

The password is hidden thanks to `terminal.ReadPassword()`. :)

**Comment by bacongobbler** [comment]
Some open questions:

- "Enter the top-level domain" is not very descriptive to a new user. Any suggestions?
- Similarly, "Enter your org". Maybe "Enter the docker org where draft will push images to (bacongobbler):"?

**Comment by bacongobbler** [comment]
0.20 has been cut 🎉 

https://github.com/kubernetes/minikube/releases/tag/v0.20.0
